workflow-ui/core-common/src/main/java/com/squareup/workflow1/ui/container/Bounds.kt

```diff
@@ -0,0 +1,52 @@
+package com.squareup.workflow1.ui.container
+
+import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
+
+/** A rectangular area. */
+@WorkflowUiExperimentalApi
+public class Bounds(
+  public val left: Int,
+  public val top: Int,
+  public val right: Int,
+  public val bottom: Int
+) {
+  public constructor() : this(0, 0, 0, 0)
+
+  public fun copy(
+    left: Int = this.left,
+    top: Int = this.top,
+    right: Int = this.right,
+    bottom: Int = this.bottom
+  ): Bounds = Bounds(left, top, right, bottom)
+
+  public operator fun component1(): Int = left
+  public operator fun component2(): Int = top
+  public operator fun component3(): Int = right
+  public operator fun component4(): Int = bottom
+
+  public val height: Int = bottom - top
+  public val width: Int = right - left
+
+  override fun equals(other: Any?): Boolean {
+    if (this === other) return true
+
+    return (other as? Bounds)?.let { otherBounds ->
+      left == otherBounds.left &&
+        top == otherBounds.top &&
+        right == otherBounds.right &&
+        bottom == otherBounds.bottom
+    } ?: false
+  }
+
+  override fun hashCode(): Int {
+    var result = left
+    result = 31 * result + top
+    result = 31 * result + right
+    result = 31 * result + bottom
+    return result
+  }
```

Comment:
Is this the reason we create the `Bounds` class and don't use `Rect`? couldn't some of this just be extensions on `Rect`? What am I missing?

Reply:
This is a java module, `workflow-ui:core-common`, not `:core-android`. We keep the built in rendering types and `ViewEnvironment` entries platform neutral.

That said, you could argue that this should just be an Android implementation detail rather than part of common. But even then, `Rect` freaks me out because it's mutable and we're passing it through a `StateFlow`. App code could stomp it.

Buuuuut…if this were in `:core-android`, I think the environment stuff could be internal. We'd only have to worry about library code using `Rect` badly, which seems managable.

WDYT? It's pretty silly for this to be anywhere but Android, isn't it?

Reply:
Gotcha. I didn't catch that nuance. I don't know what will be best but I think starting narrow with Android as it's our only use case right now makes sense.